### PR TITLE
Improve support for server rendering async routes

### LINKED ISSUE #2714: match workaround for client-side rendering doesn't work with async onEnter hooks
Hi

Following the advice of issue #2036 works to correctly render the content on the first pass until you add an async onEnter hook. When this is done the issue returns because again the location ends up being null and a null component is rendered. 

This can be reproduced using the following route:

```
export default {
  path: 'about',
  component: About,
  onEnter: function enter(nextState, replaceState, callback) {
      setTimeout(callback, 1000);
  },
}
```

I have made a small modification to the `example-react-router-server-rendering-lazy-routes` illustrating the problem here: https://github.com/ngoossens/example-react-router-server-rendering-lazy-routes

Thanks in advance for any assistance!

### PR DESCRIPTION
**Pull request by taion**
Closes #2714

### CODE REVIEW

File: modules/__tests__/_bc-serverRendering-test.js [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import expect from 'expect'
 import React, { Component } from 'react'
 import { renderToString } from 'react-dom/server'
-import match from '../match'
-import RouterContext from '../RouterContext'
+
 import Link from '../Link'
+import match from '../match'
+import RoutingContext from '../RoutingContext'
```
Comment (taion) [inline modules/__tests__/_bc-serverRendering-test.js]: This wasn't actually testing the v1 API. Now it is.

Comment (timdorr) [inline modules/__tests__/_bc-serverRendering-test.js]: Nice catch!

**Comment by timdorr** [comment]
LGTM.

**Comment by taion** [comment]
@ryanflorence 

We initially discussed not doing this in https://github.com/rackt/react-router/issues/2036#issuecomment-154563186 and subsequently, but I think this is lightweight enough, and it solves the "async `onEnter` hook" use case, and makes things mostly just work with JSPM.

What do you think?

**Comment by ngoossens** [comment]
:thumbsup:

File: modules/Router.js [hunk 2]
```diff
@@ -41,12 +41,10 @@ const Router = React.createClass({
...
   getInitialState() {
-    return {
-      location: null,
-      routes: null,
-      params: null,
-      components: null
-    }
+    // Use initial state from renderProps when available, to allow using match
+    // on client side when doing server-side rendering.
+    const { location, routes, params, components } = this.props
```
Comment (ericclemmons) [inline modules/Router.js]: Thank you for this!  I just got async routes working with RC4, had some issues, but you fixed it for me :)

:+1: @taion

**Pull request merged**